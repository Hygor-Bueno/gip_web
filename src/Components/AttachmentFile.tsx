import React, { useEffect, useState } from 'react';
import FilePreview from './FilePreview';
import { useConnection } from '../Context/ConnContext';
import { IMAGE_WEBP_QUALITY } from '../Util/Util';

function AttachmentFile(props:
  | { item_id: number; file: number; onClose?: (file: string) => void; reset?: boolean, updateAttachmentFile?: (file: string, item_id: number) => Promise<void>, fullFiles?: boolean,base64?:string } // item_id é obrigatório, onClose opcional
  | { item_id?: number; file: number; onClose: (file: string) => void; reset: boolean, updateAttachmentFile?: (file: string, item_id: number) => Promise<void>, fullFiles?: boolean,base64?:string } // onClose é obrigatório, item_id opcional
) {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [base64File, setBase64File] = useState<string>('');
  const { fetchData } = useConnection();

  const { reset } = props;
  useEffect(() => {
    if (reset) {
      setBase64File('');
    }
  }, [reset]);

  useEffect(() => {
    (async () => {
      try {
        if (props.file) {
          const req: any = await fetchData({method: "GET", params: null, pathFile: 'GTPP/TaskItem.php', urlComplement: `&id=${props.item_id}`, exception: ["no data"]})
          if (req.error) throw new Error(req.message);
          setBase64File(req.data[0]);
        }
      } catch (error: any) {
        console.error(error.message);
      }
    })();
  }, [props]);
  useEffect(() => {
    (async () => {
      props.base64 &&  setBase64File(props.base64);
    })();
  }, [props]);


  const closeModal = () => {
    setIsModalOpen(false);
    let cleanedFile = base64File;
    if(!props.fullFiles) {
      cleanedFile = base64File.split(',')[1] || '';
    }
    props.onClose?.(cleanedFile);
  };
  return (
    <div title="Anexar arquivo." onClick={() => setIsModalOpen(true)}>
      <label className="file-input-container">
        <div className={`fa fa-paperclip p-2 cursor-pointer  ${(base64File) && 'text-success shadow rounded-circle'}`}
        />
      </label>
      {isModalOpen && <AttachmentPreview item_id={props.item_id || 0} closeModal={closeModal} base64File={base64File} setBase64File={setBase64File} updateAttachmentFile={props.updateAttachmentFile} fullFiles={props.fullFiles} />}
    </div>
  );
}

function AttachmentPreview(props: { closeModal: () => void; item_id: number, base64File: string, setBase64File: (value: string) => void, updateAttachmentFile?: (file: string, item_id: number) => Promise<void>, fullFiles?: boolean }) {
  const { base64File, setBase64File, closeModal, item_id } = props;

    function handleFileChange(event: any) {
      const file = event.target.files[0];
      if (file) {
        const reader = new FileReader();
        reader.onload = (e) => {
          const img = new Image();
          img.onload = () => {
            const canvas = document.createElement("canvas");
            canvas.width = img.width;
            canvas.height = img.height;
            const ctx = canvas.getContext("2d");
            ctx?.drawImage(img, 0, 0);
            const webpData = canvas.toDataURL("image/webp", IMAGE_WEBP_QUALITY); // qualidade ajustável
            setBase64File(webpData);
          };
          img.src = e.target?.result?.toString() || '';
        };
        reader.readAsDataURL(file);
      }
    }


  
    useEffect(() => {
    function handlePaste(event: ClipboardEvent) {
      const items = event.clipboardData?.items;
      if (!items) return;
      for (const item of items) {
        if (item.type.indexOf("image") === 0) {
          const file = item.getAsFile();
          if (file) {
            const reader = new FileReader();
            reader.onload = (e) => {
              const img = new Image();
              img.onload = () => {
                const canvas = document.createElement("canvas");
                canvas.width = img.width;
                canvas.height = img.height;
                const ctx = canvas.getContext("2d");
                ctx?.drawImage(img, 0, 0);
                let webpData = canvas.toDataURL("image/webp", IMAGE_WEBP_QUALITY);
                if (!webpData.startsWith("data:image/webp")) {
                  console.log("Conversão para WebP falhou, usando PNG como fallback.");
                  webpData = canvas.toDataURL("image/png");
                }

                setBase64File(webpData);
              };
              img.src = e.target?.result?.toString() || '';
            };
            reader.readAsDataURL(file);
          }
        }
      }
    }

    window.addEventListener("paste", handlePaste as EventListener);
    return () => {
      window.removeEventListener("paste", handlePaste as EventListener);
    };
  }, [setBase64File])

  return (
    <div className="modal-overlay" onClick={closeModal}>
      <div style={{ maxWidth: "75%", maxHeight: "90%" }} className="d-flex flex-column align-items-center bg-white p-4 rounded" onClick={(e) => e.stopPropagation()}>
        <div className='d-flex align-items-center justify-content-between w-100'>
          <span className='h5'>Anexo: </span>
          {base64File && <button title="Remover anexo." style={{ minWidth: "25%" }} onClick={async () => {
            if (props.updateAttachmentFile) {
              await props.updateAttachmentFile('', item_id);
            }
            setBase64File('');
          }} className="btn btn-danger m-2 fa-solid fa-trash" />}
        </div>
        <div className='d-flex flex-column align-items-center w-100 h-100 overflow-auto'>
          {base64File ?
            <FilePreview base64File={base64File}/>
            :
            <>
              <label style={{ minHeight: "60px", height: "4vw", minWidth: "60px", width: "4vw" }} className='d-flex justify-content-center align-items-center btn btn-outline-primary text-primary fa fa-paperclip' >
                <input
                  type="file"
                  onChange={handleFileChange}
                  style={{ display: 'none' }}
                />
              </label>
              <label className='text-danger text-bold' style={{fontSize: '0.950rem', width: '15rem', textAlign: 'center', marginTop: '1rem'}}>Pode usar CTRL+C e V para colar a imagem ou anexar ela clicando no botão</label>
            </>
          }
        </div>

        <div className='d-flex align-items-center justify-content-around w-100'>
          <button title={"Salvar arquivo"} disabled={!item_id} style={{ minWidth: "25%" }} onClick={async () => {
            if (props.updateAttachmentFile) {
              await props.updateAttachmentFile(base64File.replace(/^data:image\/\w+;base64,/, ""), item_id);
            }
            closeModal();
          }} className="btn btn-success m-2">
            Salvar
          </button>
          <button title={"Voltar"} style={{ minWidth: "25%" }} onClick={closeModal} className="btn btn-danger m-2">
            Voltar
          </button>
        </div>

      </div>
    </div>
  );

}

export default AttachmentFile;
