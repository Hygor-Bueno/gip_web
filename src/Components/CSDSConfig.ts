export const fieldsetsCSDS = (
    optionsCompay: [{ value: string, label: string }],
    onCompany: (id: number) => void,
    optionsDepartament: [{ value: string, label: string }],
    onDepartament: (id: number) => void,
    optionsStore: [{ value: string, label: string }],
    onSubDepartament: (id: number) => void,
    optionsSubDepartament: [{ value: string, label: string }],
    selectSubDep: (id: number) => void,
    nameUser: string,
    setNameUser: (value: string) => void,
    selectCompany: string
) => [
        {
            attributes: { className: 'ps-2 col-3' },
            item: {
                label: 'Nome',
                captureValue: {
                    onChange: (e: any) => { setNameUser(e.target.value) },
                    type: 'text',
                    value: nameUser,
                    placeholder: 'Nome do usuÃ¡rio',
                    name: 'user_name',
                    className: 'form-control'
                }
            }
        }, {
            attributes: { className: 'ps-2 col-2 overflow-hidden', },
            item: {
                label: 'Companhias',
                captureValue: {
                    type: 'select',
                    value: selectCompany,
                    onChange: (e: any) => { onCompany(parseInt(e.target.value)) },
                    placeholder: '',
                    name: 'company_id',
                    className: 'form-control',
                    required: true,
                    id: '',
                    options: optionsCompay
                },
            }
        }, {
            attributes: { className: 'ps-2 col-2 overflow-hidden', },
            item: {
                label: 'Lojas',
                captureValue: {
                    type: 'select',
                    onChange: (e: any) => { onDepartament(parseInt(e.target.value)) },
                    placeholder: '',
                    name: 'shop_id',
                    className: 'form-control',
                    required: true,
                    id: '',
                    options: optionsDepartament
                },
            }
        }
        , {
            attributes: { className: 'ps-2 col-2 overflow-hidden', },
            item: {
                label: 'Depart.',
                captureValue: {
                    type: 'select',
                    onChange: (e: any) => { onSubDepartament(parseInt(e.target.value)) },
                    placeholder: '',
                    name: 'depart_id',
                    className: 'form-control',
                    required: true,
                    id: '',
                    options: optionsStore
                },
            }
        }
        , {
            attributes: { className: 'ps-2 col-3 overflow-hidden', },
            item: {
                label: 'Subdep.',
                captureValue: {
                    type: 'select',
                    onChange: (e: any) => { selectSubDep(parseInt(e.target.value)) },
                    placeholder: '',
                    name: 'depart_id',
                    className: 'form-control',
                    required: true,
                    id: '',
                    options: optionsSubDepartament
                },
            }
        }
    ];