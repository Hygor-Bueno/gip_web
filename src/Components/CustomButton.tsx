import React from 'react';
type CustomButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
    value?: string; 
    customStyle?:string;
};

export function CustomButton({ value,customStyle, children, ...props }: CustomButtonProps) {
    return (
        <button className={customStyle} {...props}>
            {value ? value : children}
        </button>
    );
}