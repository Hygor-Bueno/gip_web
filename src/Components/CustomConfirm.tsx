import React from 'react';

export default function ConfirmModal({
  title,
  message,
  onConfirm,
  onClose,
}: {
  title: string;
  message: string;
  onConfirm: () => void;
  onClose: () => void;
}): JSX.Element {
  return (
    <>
      <style>{`
        .confirm-modal-overlay {
          position: fixed;
          top: 0;
          left: 0;
          width: 100vw;
          height: 100dvh;
          background-color: rgba(0, 0, 0, 0.5);
          display: flex;
          justify-content: center;
          align-items: center;
          z-index: 1050;
          padding: 1rem;
        }

        .confirm-modal-container {
          background-color: #fff;
          border-radius: 0.5rem;
          box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
          width: 100%;
          max-width: 30rem;
          overflow: hidden;
          display: flex;
          flex-direction: column;
        }

        .confirm-modal-header {
          padding: 1rem;
          border-bottom: 1px solid #dee2e6;
          display: flex;
          justify-content: space-between;
          align-items: center;
        }

        .confirm-modal-title {
          margin: 0;
          font-size: 1.25rem;
        }

        .confirm-modal-body {
          padding: 1rem;
          font-size: 1rem;
        }

        .confirm-modal-body p {
          margin: 0;
        }

        .confirm-modal-footer {
          padding: 1rem;
          border-top: 1px solid #dee2e6;
          display: flex;
          justify-content: flex-end;
          gap: 0.5rem;
        }
      `}</style>

      <div className="confirm-modal-overlay">
        <div className="confirm-modal-container">
          <div className="confirm-modal-header">
            <h5 className="confirm-modal-title">{title}</h5>
            <button className="btn-close" onClick={onClose} aria-label="Close"></button>
          </div>

          <div className="confirm-modal-body">
            <p>{message}</p>
          </div>

          <div className="confirm-modal-footer">
            <button title="Cancelar ação." className="btn btn-secondary" onClick={onClose}>
              Cancelar
            </button>
            <button title="Confirmar ação" className="btn btn-primary" onClick={onConfirm}>
              Confirmar
            </button>
          </div>
        </div>
      </div>
    </>
  );
}
