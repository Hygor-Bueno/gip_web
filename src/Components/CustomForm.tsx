import React from 'react';

type CustomFormProps = React.FormHTMLAttributes<HTMLFormElement> & {
  fieldsets: {
    attributes?: React.FieldsetHTMLAttributes<HTMLFieldSetElement>;
    item: {
      label: string;
      mandatory?: boolean;
      captureValue:
        | React.InputHTMLAttributes<HTMLInputElement>
        | (React.SelectHTMLAttributes<HTMLSelectElement> & { options?: SelectOption[] })  // Aqui adicionamos options somente para select
        | React.TextareaHTMLAttributes<HTMLTextAreaElement>;
    };
    legend?: {
      style?: string;
      text?: string;
    };
    buttons?: [];
    
  }[];
  onAction?: any;
  titleButton?: any;
};

interface SelectOption {
  value: string | number;
  label: string;
}

function CustomForm({ fieldsets, titleButton="Login", ...formProps }: CustomFormProps) {
  return (
    <form {...formProps}>
      {fieldsets.map((fieldset, fieldsetIndex) => (
        <fieldset key={fieldsetIndex} {...fieldset.attributes}>
          <legend className={fieldset.legend?.style}>{fieldset.legend?.text}</legend>
          <label>
            {fieldset.item.label}
            <b className={fieldset.item.mandatory ? 'text-danger' : ''}>
              {fieldset.item.mandatory ? ' *' : ''}
            </b>
            :
          </label>
          {renderField(fieldset.item.captureValue)}
        </fieldset>
      ))}
      {/* <button className='btn mt-5 my-2'>Enviar</button> */}
      <button className="btn my-2">{titleButton}</button>
    </form>
  );
}

function renderField(
  captureValue:
    | React.InputHTMLAttributes<HTMLInputElement>
    | (React.SelectHTMLAttributes<HTMLSelectElement> & { options?: SelectOption[] }) // Certifique-se que options s√≥ existe em select
    | React.TextareaHTMLAttributes<HTMLTextAreaElement>
) {
  if ('type' in captureValue) {
    switch (captureValue.type) {
      case 'select':
        return (
          <SelectField
            {...(captureValue as React.SelectHTMLAttributes<HTMLSelectElement> & { options?: SelectOption[] })}
            // @ts-ignore
            options={captureValue?.options || []} 
          />
        );
      case 'textarea':
        return <TextareaField {...(captureValue as React.TextareaHTMLAttributes<HTMLTextAreaElement>)} />;
      default:
        return <InputField {...(captureValue as React.InputHTMLAttributes<HTMLInputElement>)} />;
    }
  }
  return <InputField {...(captureValue as React.InputHTMLAttributes<HTMLInputElement>)} />;
}

export function InputField(props: React.InputHTMLAttributes<HTMLInputElement>) {
  return <input {...props} />;
}

export function SelectField(
  props: React.SelectHTMLAttributes<HTMLSelectElement> & { options: SelectOption[] }
) {
  return (
    <select {...props} className={`form-control ${props.className}`}>
      {props.options.map((option, index) => (
        <option key={index} value={option.value}>
          {option.label}
        </option>
      ))}
    </select>
  );
}

export function SelectFieldDefault (props: {
  label?: string;
  value?: string | number;
  onChange?: any;
  className?: string;
  options: {label?: string, value?: string | number}[];
}) {
  return (
    <label>
      {props.label}
      <select value={props.value} onChange={props.onChange} className={`form-select ${props.className}`}>
        <option defaultValue={""} value={""}>Selecione</option>
        {props.options.map(({label, value}, key: number) => (
          <option key={`opt_${key}`} value={value}>{label}</option>
        ))}
      </select>
    </label>
  )
}


export function InputCheckbox(props: {label: string, value: boolean , onChange: any}) {
  return (
    <label className='cursor-pointer'>
      <input type="checkbox" checked={props.value} onChange={props.onChange} className='form-check-input' />{" "}
      {props.label}
    </label>
  )
}

export function TextareaField(props: React.TextareaHTMLAttributes<HTMLTextAreaElement>) {
  return <textarea {...props} />;
}

export default CustomForm;
