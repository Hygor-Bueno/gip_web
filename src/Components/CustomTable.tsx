import React, { useEffect, useState } from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import { tItemTable } from "../types/types";
const defaultImage = require('../Assets/Image/groupCLPP.png');

export default function TableComponent(props: { 
  list: tItemTable[], 
  onConfirmList: (selected: tItemTable[]) => void, 
  selectedItems?: tItemTable[],
  maxSelection?: number
}) {
  const [sortConfig, setSortConfig] = useState<{ key: string; direction: string } | null>(null);
  const [filters, setFilters] = useState<{ [key: string]: string }>({});
  const [activeFilter, setActiveFilter] = useState<string | null>(null);
  const [selectedRows, setSelectedRows] = useState<tItemTable[]>(props.selectedItems || []);

  // Atualiza o estado selectedRows quando props.selectedItems mudar
  useEffect(() => {
    if (props.selectedItems) {
      setSelectedRows(props.selectedItems);
    }
  }, [props.selectedItems]);
  console.log(selectedRows)

  const columnKeys = Object.keys(props.list[0]);
  const columnHeaders = columnKeys.reduce((acc, key) => {
    acc[key] = props.list[0][key].tag;
    return acc;
  }, {} as { [key: string]: string });

  const sortedItemTable = [...props.list].sort((a, b) => {
    if (!sortConfig) return 0;
    const { key, direction } = sortConfig;
    const valueA = a[key].value.toLowerCase();
    const valueB = b[key].value.toLowerCase();
    return direction === "asc" ? valueA.localeCompare(valueB) : valueB.localeCompare(valueA);
  });

  const filteredItemsTable = sortedItemTable.filter((item) =>
    Object.keys(filters).every((key) =>
      item[key].value.toLowerCase().includes(filters[key].toLowerCase())
    )
  );

  function handleSort(key: string) {
    setSortConfig((prev) => {
      if (prev && prev.key === key) {
        return { key, direction: prev.direction === "asc" ? "desc" : "asc" };
      }
      return { key, direction: "asc" };
    });
  }

  function handleFilterChange(key: string, value: string) {
    setFilters((prev) => ({ ...prev, [key]: value }));
  }

  function toggleRowSelection(item: tItemTable) {
    setSelectedRows((prev) => {
      if (prev.includes(item)) {
        return prev.filter((row) => row !== item);
      } else {
        if (props.maxSelection && prev.length >= props.maxSelection) {
          return prev; // Não adiciona mais itens se o limite for atingido
        }
        return [...prev, item];
      }
    });
  }

  function RenderCell(props: { value: string; isImage?: boolean }) {
    return props.isImage ? (
      <img className="photoCircle rounded-circle" src={props.value ? `data:image/png;base64,${props.value}` : defaultImage} />
    ) : (
      <span>{props.value}</span>
    );
  }

  return (
    <div className="d-flex flex-column w-100 h-100 p-3">
      <div className="overflow-auto">
        <table className="table table-bordered table-striped">
          <thead className="table-light">
            <tr>
              {columnKeys.filter((key) => !props.list[0][key].ocultColumn).map((key) => (
                <th key={key} className="position-relative">
                  <div style={{ minWidth: props.list[0][key].minWidth ? props.list[0][key].minWidth : "auto" }} className="d-flex justify-content-between align-items-center">
                    {activeFilter === key ? (
                      <input
                        type="text"
                        className="form-control"
                        placeholder="Filtrar"
                        autoFocus
                        value={filters[key] || ""}
                        onChange={(e) => handleFilterChange(key, e.target.value)}
                        onBlur={() => setActiveFilter(null)}
                      />
                    ) : (
                      <span onClick={() => handleSort(key)}>
                        {columnHeaders[key]} {sortConfig?.key === key ? (sortConfig.direction === "asc" ? "▲" : "▼") : ""}
                      </span>
                    )}
                    <button className="btn btn-sm btn-light ms-2" onClick={(e) => { e.stopPropagation(); setActiveFilter(activeFilter === key ? null : key); }}>
                      <i className="fas fa-search"></i>
                    </button>
                  </div>
                </th>
              ))}
            </tr>
          </thead>
          <tbody>
            {filteredItemsTable.map((item, index) => (
              <tr key={index} className={selectedRows.includes(item) ? "table-success" : ""} onClick={() => toggleRowSelection(item)}>
                {columnKeys.map((key) => !item[key].ocultColumn && (
                  <td key={key} className="py-2">
                    <RenderCell isImage={item[key].isImage} value={item[key]?.value} />
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
      <div className="w-100 d-flex justify-content-around">
        <button title="Confirmar seleção atual" className="btn btn-primary mt-3" onClick={() => props.onConfirmList(selectedRows)}>
          Confirmar Seleção
        </button>
        <button title="Limpar seleção atual" className="btn btn-secondary text-white mt-3" onClick={() => setSelectedRows([])}>
          Limpar Seleção
        </button>
      </div>
    </div>
  );
};