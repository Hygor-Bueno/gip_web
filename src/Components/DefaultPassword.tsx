import React, { useState } from "react";
import { formChangePassword } from "../Configs/DefaultPassword";
import CustomForm from "./CustomForm";
import { useMyContext } from "../Context/MainContext";
import { Connection } from "../Connection/Connection";

export default function DefaultPassword(props: {
    open: boolean,
    user: {
        login: string
        password: string
    },
    onClose:()=>void
}): JSX.Element {
    const [password, setPassword] = useState<string>("");
    const [confirm, setConfirm] = useState<string>("");

    const { setLoading } = useMyContext();
    async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
        try {
            setLoading(true);
            event.preventDefault();
            if (password !== confirm) throw new Error("As senhas n√£o conferem");
            const connection = new Connection("15",true);
            const req:any = await connection.put({ user: props.user.login, password: props.user.password, new_password:password, confirm_password:confirm}, "CCPP/Login.php");
            if(req.error) throw new Error(req.message);
            alert("Senha alterada com sucesso!");
            props.onClose();
        } catch (error: any) {
            alert(error.message);
        } finally {
            setLoading(false);
        }
    };
    return props.open ? (
        <div onClick={()=>props.onClose()} className='d-flex  align-items-center justify-content-center bg-black bg-opacity-25 position-absolute h-100 w-100 top-0'>
            <div className='d-flex flex-column bg-white col-8 col-sm-6 col-md-4 col-lg-3  py-4 rounded'>
                <header className='d-flex aligm-itens-center justify-content-center my-4  w-100'>
                    <h1>Hora de trocar sua senha!</h1>
                </header>
                <CustomForm
                    classButton='btn btn-success'
                    onSubmit={handleSubmit}
                    titleButton={"Confirmar"}
                    className='d-flex flex-column align-items-center justify-content-center h-100'
                    fieldsets={
                        formChangePassword(
                            (e: any) => setPassword(e.target.value),
                            (e: any) => setConfirm(e.target.value),
                            password,
                            confirm
                        )
                    }
                />
            </div>
        </div>
    ) : <React.Fragment />;
}