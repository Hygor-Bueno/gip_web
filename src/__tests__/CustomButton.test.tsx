import { render, screen } from "@testing-library/react";
import { CustomButton, InputCheckButton } from "../Components/CustomButton";
import userEvent from "@testing-library/user-event";

test('renderiza o botão e verifica se o botão tem um texto', () => {
    render(<CustomButton children="Clique aqui!" />);
    expect(screen.getByText('Clique aqui!')).toBeInTheDocument();
});

test('Chama a função ao clicar no botão', async () => {
    const mockclick = jest.fn();
    render(<CustomButton onClick={mockclick} children="Clique aqui!" />);
    const btn = screen.getByText("Clique aqui!");
    await userEvent.click(btn);
    expect(mockclick).toHaveBeenCalledTimes(1);
});

test('Verifica se consigo colocar uma className nesse componente', () => {
    render(<CustomButton customStyle="have-style" children="Clique aqui!" />);
    const btn = screen.getByText("Clique aqui!");
    expect(btn).toHaveClass('have-style');
});

test('Chama o botão checkbox e verifica se ele renderiza na tela', () => {
    const mockAction = jest.fn();
    render(<InputCheckButton inputId="name-test" onAction={mockAction} nameButton="Testing-checkbox" />);
});

test('Chama o botão checkbox e verifica se tem um nome nele', () => {
  const mockAction = jest.fn();

  render(
    <InputCheckButton
      inputId="name-test"
      labelText="Clique aqui!"
      onAction={mockAction}
      nameButton="Testing-checkbox"
    />
  );

  const btnCheck = screen.getByRole("button");

  expect(btnCheck).toHaveAttribute("title", "Testing-checkbox");
});

test('Clica na label e ativa o checkbox, chamando onAction', async () => {
    const mockAction = jest.fn();
    const user = userEvent.setup();

    render(
        <InputCheckButton
        inputId="name-test"
        labelText="Clique aqui!"
        onAction={mockAction}
        nameButton="Testing-checkbox"
        />
    );

    const label = screen.getByText('Clique aqui!');
    await user.click(label);
    expect(mockAction).toHaveBeenCalledWith(true);
});

test('Clica 2 vezes na label e altera valor do checkbox', async () => {
    const mockAction = jest.fn();
    const user = userEvent.setup();

    render (
        <InputCheckButton
            inputId="name-test"
            labelText="Clique aqui!"
            onAction={mockAction}
            nameButton="Testing-checkbox"
        />
    );

    const label = screen.getByText('Clique aqui!');

    await user.click(label);
    await user.click(label);

    expect(mockAction).toHaveBeenNthCalledWith(1, true);
    expect(mockAction).toHaveBeenNthCalledWith(2, false);
});

test('Verifica se o checkbox muda o estado visualmente', async () => {
    const mockAction = jest.fn();
    const user = userEvent.setup();

    render(
        <InputCheckButton
            inputId="name-test"
            labelText="Clique aqui!"
            onAction={mockAction}
            nameButton="Testing-checkbox"
        />
    );

    const checkbox = screen.getByLabelText('Clique aqui!');
    expect(checkbox).not.toBeChecked();


    await user.click(screen.getByText('Clique aqui!'));

    expect(checkbox).toBeChecked();
});

test('Verifica se o checkbox está desativado visualmente', async () => {
    const mockAction = jest.fn(); 

    render(
        <InputCheckButton
            inputId="name-test"
            labelText="Clique aqui!"
            onAction={mockAction}
            nameButton="Testing-checkbox"
        />
    );

    const labelckeckbox = screen.getByLabelText('Clique aqui!');
    expect(labelckeckbox).not.toBeChecked();
});

test('muda o icone ao alterar o checkbox', async () => {
    const user = userEvent.setup();

    render(
        <InputCheckButton
            inputId="name-test"
            labelText="icone!"
            onAction={jest.fn()}
            labelIconConditional={['icon-checked', 'icon-unchecked']}
            nameButton="Botão com ícone"
        />
    );

    const label = screen.getByText('icone!');

    expect(label).toHaveClass('icon-unchecked');

    await user.click(label);

    expect(label).toHaveClass('icon-checked');
});

test('muda o ícone estático se fornecido', () => {
    render(
        <InputCheckButton
            inputId="name-test"
            labelText="icone!"
            onAction={jest.fn()}
            labelIcon="icon-fixed"
            nameButton="Botão com ícone"
        />
    );

    const label = screen.getByText('icone!');
    expect(label).toHaveClass('icon-fixed');
});

test('altera cor do botão quando highlight está ativado e checkbox marcado', async () => {
    const user = userEvent.setup();
    render(
        <InputCheckButton
            inputId="name-test"
            labelText="Highlight test"
            onAction={jest.fn()}
            highlight={true}
            labelIcon="icon-fixed"
            nameButton="Botão com ícone"
        />
    );

    const label = screen.getByText('Highlight test');
    const button = screen.getByRole('button');

    expect(button).toHaveClass('btn-light');

    await user.click(label);

    expect(button).toHaveClass('btn-success');
});

test('aplica cor do texto definida por labelColor', () => {
  render(
    <InputCheckButton
      inputId="checkbox-color"
      labelText="Colorido"
      onAction={jest.fn()}
      labelColor="text-danger"
      nameButton="Botão"
    />
  );

  const label = screen.getByText('Colorido');
  expect(label).toHaveClass('text-danger');
});

test('não quebra se onAction lançar erro', async () => {
  const user = userEvent.setup();
  const mockAction = jest.fn().mockRejectedValue(new Error('Falha'));

  render(
    <InputCheckButton
      inputId="checkbox-error"
      labelText="Erro"
      onAction={mockAction}
      nameButton="Botão"
    />
  );

  const label = screen.getByText('Erro');
  await user.click(label);

  expect(mockAction).toHaveBeenCalled();
});
