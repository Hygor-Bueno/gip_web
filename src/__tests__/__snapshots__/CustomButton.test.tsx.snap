// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`O button deve corresponder ao snapshot visual 1`] = `
<DocumentFragment>
  <button
    class="btn btn-success"
    title="Botão customizado"
  >
    Clique aqui!
  </button>
</DocumentFragment>
`;

exports[`O input checkbox deve corresponder ao snapshot visual 1`] = `
<DocumentFragment>
  <button
    class="btn btn-light p-0"
    title="Botão"
    type="button"
  >
    <input
      class="cursor-pointer"
      hidden=""
      id="checkbox-error"
      type="checkbox"
    />
    <label
      class="cursor-pointer  text-dark  p-2"
      for="checkbox-error"
    >
      Erro
    </label>
  </button>
</DocumentFragment>
`;
